### PR TITLE
Vision: Add gRPC support for face detection.

### PR DESCRIPTION
**Pull request by daspecster**
~Based from #2918~

This adds gRPC support for `face_detect()`.

### CODE REVIEW

File: system_tests/vision.py [hunk 1]
```diff
@@ -63,7 +63,6 @@ def _assert_coordinate(self, coordinate):
         if coordinate is None:
             return
         self.assertIsInstance(coordinate, (int, float))
-        self.assertNotEqual(coordinate, 0.0)
```
Comment (dhermes) [inline system_tests/vision.py]: Why not?

File: system_tests/vision.py [hunk 2]
```diff
@@ -133,6 +132,7 @@ def test_detect_logos_gcs(self):
 
 class TestVisionClientFace(BaseVisionTestCase):
     def setUp(self):
+        Config.CLIENT = vision.Client(use_gax=True)
```
Comment (dhermes) [inline system_tests/vision.py]: Why?

File: system_tests/vision.py [hunk 3]
```diff
@@ -423,6 +425,7 @@ def _assert_text(self, text):
 
     def test_detect_text_content(self):
         client = Config.CLIENT
+        client._use_gax = True
```
Comment (dhermes) [inline system_tests/vision.py]: I am :-1: on this. See my comments on the other PR

File: vision/google/cloud/vision/face.py [hunk 4]
```diff
@@ -22,6 +22,29 @@
...
+def _get_pb_likelihood(likelihood):
+    """Convert gRPC Likelihood integer value to Likelihood instance.
+
+    :type likelihood: int
+    :param likelihood: Protobuf integer representing ``Likelihood``.
+
+    :rtype: :class:`~google.cloud.vision.likelihood.Likelihood`
+    :returns: Instance of ``Likelihood`` converted from gRPC value.
+    """
+    likelihood_pb = LIKELIHOOD_PB.get(likelihood)
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Where does `likelihood` come from the `pb2` module should have a way to get the enum

File: vision/google/cloud/vision/face.py [hunk 5]
```diff
@@ -42,6 +65,23 @@ def from_api_repr(cls, response):
 
         return cls(roll, pan, tilt)
 
+    @classmethod
+    def from_pb(cls, response):
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Similar comment here as in #2918, the fact that the argument is a `response` is less important than the fact this it's an `Angle`.

File: vision/google/cloud/vision/face.py [hunk 6]
```diff
@@ -42,6 +65,23 @@ def from_api_repr(cls, response):
...
+        """Factory: convert gRPC Angle object to local Angle object.
+
+        :type response: :class:`~google.cloud.grpc.vision.v1.\
+                        image_annotator_pb2.FaceAnnotation`
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Copy-pasta?

File: vision/google/cloud/vision/face.py [hunk 7]
```diff
@@ -101,6 +141,24 @@ def from_api_repr(cls, response):
         return cls(joy_likelihood, sorrow_likelihood, surprise_likelihood,
                    anger_likelihood)
 
+    @classmethod
+    def from_pb(cls, response):
+        """Factory: construct `Emotions` from Vision API response.
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Single-backtick: Markdown
Double-backtick: RST

File: vision/google/cloud/vision/face.py [hunk 8]
```diff
@@ -101,6 +141,24 @@ def from_api_repr(cls, response):
...
+
+        :type response: dict
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: ISTM that `response` isn't a dict?

File: vision/google/cloud/vision/face.py [hunk 9]
```diff
@@ -175,13 +233,39 @@ def from_api_repr(cls, response):
...
+        fd_bounds = FDBounds.from_pb(response.fd_bounding_poly)
+        headwear_likelihood = Likelihood[
+            LIKELIHOOD_PB[response.headwear_likelihood]]
+        image_properties = FaceImageProperties.from_pb(response)
+        landmarks = Landmarks.from_pb(response.landmarks)
+        landmarking_confidence = response.landmarking_confidence
+
+        return cls(angles, bounds, detection_confidence, emotions, fd_bounds,
+                   headwear_likelihood, image_properties, landmarks,
+                   landmarking_confidence)
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Whoa! A mouthful here.

File: vision/google/cloud/vision/face.py [hunk 10]
```diff
@@ -326,6 +410,18 @@ def from_api_repr(cls, response):
 
         return cls(blurred, underexposed)
 
+    @classmethod
+    def from_pb(cls, response):
+        """Factory: construct image properties from image.
+
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Need args description

File: vision/google/cloud/vision/face.py [hunk 11]
```diff
@@ -353,41 +449,110 @@ class LandmarkTypes(Enum):
...
-    LEFT_EAR_TRAGION = 'LEFT_EAR_TRAGION'
-    RIGHT_EAR_TRAGION = 'RIGHT_EAR_TRAGION'
-    LEFT_EYE_PUPIL = 'LEFT_EYE_PUPIL'
-    RIGHT_EYE_PUPIL = 'RIGHT_EYE_PUPIL'
-    FOREHEAD_GLABELLA = 'FOREHEAD_GLABELLA'
-    CHIN_GNATHION = 'CHIN_GNATHION'
-    CHIN_LEFT_GONION = 'CHIN_LEFT_GONION'
-    CHIN_RIGHT_GONION = 'CHIN_RIGHT_GONION'
+    # Unknown face landmark detected. Should not be filled.
+    UNKNOWN_LANDMARK = 0
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Don't hardcode this, it's in the generated `pb2` module

File: vision/google/cloud/vision/face.py [hunk 12]
```diff
@@ -414,6 +579,20 @@ def from_api_repr(cls, response_landmark):
         landmark_type = LandmarkTypes[response_landmark['type']]
         return cls(position, landmark_type)
 
+    @classmethod
+    def from_pb(cls, response_landmark):
+        """Factory: construct an instance of a Landmark from a response.
+
+        :type response_landmark: dict
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Not a dict?

File: vision/google/cloud/vision/face.py [hunk 13]
```diff
@@ -434,8 +613,42 @@ def landmark_type(self):
...
+
+        :type response: dict
+        :param response: JSON face annotation.
+
+        :rtype: :class:`~google.cloud.vision.face.Landmarks`
+        :returns: Instance of ``Landmarks`` populated with facial landmarks.
+        """
+        landmarks = []
+        for landmark in response:
+            landmarks.append(Landmark.from_api_repr(landmark))
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: Why not a list comprehension?

File: vision/google/cloud/vision/image.py [hunk 14]
```diff
@@ -54,7 +54,7 @@ def __init__(self, client, content=None, filename=None, source_uri=None):
                 content = file_obj.read()
 
         if content is not None:
-            content = _bytes_to_unicode(b64encode(_to_bytes(content)))
+            content = _to_bytes(content)
```
Comment (dhermes) [inline vision/google/cloud/vision/image.py]: Why?

File: vision/google/cloud/vision/image.py [hunk 15]
```diff
@@ -67,7 +67,7 @@ def as_dict(self):
         """
         if self.content:
             return {
-                'content': self.content
+                'content': _bytes_to_unicode(b64encode(self.content))
```
Comment (dhermes) [inline vision/google/cloud/vision/image.py]: Why? (Same as above, I assume?) Also, trailing comma

File: vision/unit_tests/test_annotations.py [hunk 16]
```diff
@@ -88,6 +88,21 @@ def test_make_entity_from_pb(self):
         self.assertEqual(entity.locations[0].longitude, 2.0)
 
 
+class TestMakeFacesFromPB(unittest.TestCase):
```
Comment (dhermes) [inline vision/unit_tests/test_annotations.py]: Please follow our style: `Test__make_faces_from_pb`

File: system_tests/vision.py (hunk 1 above)
Comment (daspecster) [inline system_tests/vision.py]: So with a face, there could be a facial feature that starts at coordinate 0,0 of a facial bounds box which would be 0,0.

File: vision/google/cloud/vision/face.py (hunk 4 above)
Comment (daspecster) [inline vision/google/cloud/vision/face.py]: Wouldn't I need to make a `face.py` and a `face_pb.py` file in that case? So that the imports don't crash if protobuf is disabled/not available?

File: system_tests/vision.py (hunk 1 above)
Comment (dhermes) [inline system_tests/vision.py]: So why was it `0.0` before?

File: vision/google/cloud/vision/face.py (hunk 4 above)
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: I think the generated stuff `try/except`-s the `grpc` imports now. Double check?

File: system_tests/vision.py (hunk 1 above)
Comment (daspecster) [inline system_tests/vision.py]: Sorry wrong reason. I forgot, protobuf was returning an `int` there so I was getting `AssertionError: 0 == 0.0`.
I could do `self.assertNotEqual(0,[0, 0.0])` I suppose.

Comment (dhermes) [inline system_tests/vision.py]: In which version of Python? They are meant to be equal.

**Comment by daspecster** [comment]
Rebased on master.

File: vision/google/cloud/vision/image.py (hunk 14 above)
Comment (daspecster) [inline vision/google/cloud/vision/image.py]: So the PB version of base64 encodes the data for us.

So for the PB version I need the content unencoded and for the HTTP version I need the data encoded.

File: vision/google/cloud/vision/face.py (hunk 11 above)
Comment (daspecster) [inline vision/google/cloud/vision/face.py]: So I converted the HTTP version to handle it with integers so that both PB and HTTP versions handle the types the same way.

I might be able to change the HTTP version so it just uses the values that are returned but that seems kind of like the wild west.

File: vision/google/cloud/vision/image.py (hunk 15 above)
Comment (daspecster) [inline vision/google/cloud/vision/image.py]: Yup same as above. HTTP needs the content to be base64 encoded and the PB version needs the content not to be.

File: system_tests/vision.py [hunk 17]
```diff
@@ -137,6 +137,7 @@ def test_detect_logos_gcs(self):
 
 class TestVisionClientFace(BaseVisionTestCase):
     def setUp(self):
+        Config.CLIENT = vision.Client()
```
Comment (dhermes) [inline system_tests/vision.py]: Do this in `setupModule` and nowhere else.

File: vision/google/cloud/vision/_gax.py [hunk 18]
```diff
@@ -85,7 +83,7 @@ def _to_gapic_image(image):
               :class:`~google.cloud.vision.image.Image`.
     """
     if image.content is not None:
-        return image_annotator_pb2.Image(content=_to_bytes(image.content))
+        return image_annotator_pb2.Image(content=image.content)
```
Comment (dhermes) [inline vision/google/cloud/vision/_gax.py]: Why did this drop?

File: vision/google/cloud/vision/client.py [hunk 19]
```diff
@@ -58,7 +58,7 @@ class Client(JSONClient):
     _vision_api_internal = None
 
     def __init__(self, project=None, credentials=None, http=None,
-                 use_gax=False):
+                 use_gax=None):
```
Comment (dhermes) [inline vision/google/cloud/vision/client.py]: Why this change?

File: vision/google/cloud/vision/face.py [hunk 20]
```diff
@@ -84,19 +119,38 @@ def __init__(self, joy_likelihood, sorrow_likelihood,
...
+        :type emotions: dict
+        :param emotions: Response dictionary representing a face.
+
+        :rtype: :class:`~google.cloud.vision.face.Emotions`
+        :returns: Populated instance of ``Emotions``.
+        """
+        joy_likelihood = Likelihood[emotions['joyLikelihood']]
+        sorrow_likelihood = Likelihood[emotions['sorrowLikelihood']]
+        surprise_likelihood = Likelihood[emotions['surpriseLikelihood']]
+        anger_likelihood = Likelihood[emotions['angerLikelihood']]
```
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: `Likelihood` is an `enum.Enum`?

File: vision/google/cloud/vision/face.py (hunk 11 above)
Comment (dhermes) [inline vision/google/cloud/vision/face.py]: I'm not following. Are you saying it **needs** to be hard-coded?

File: vision/google/cloud/vision/geometry.py [hunk 21]
```diff
@@ -219,7 +231,8 @@ def x_coordinate(self):
         :rtype: float
         :returns: X position coordinate.
         """
-        return self._x_coordinate
+        if self._x_coordinate != 0:
+            return self._x_coordinate
```
Comment (dhermes) [inline vision/google/cloud/vision/geometry.py]: No need for this branch if you don't store `0` in the constructor (i.e. just use `None`)

File: vision/google/cloud/vision/image.py (hunk 14 above)
Comment (dhermes) [inline vision/google/cloud/vision/image.py]: WAT? It's an image, it should always be bytes? If `base64.b64encode` requires a `str` in Python 3, just convert at the call site.

File: vision/google/cloud/vision/image.py (hunk 15 above)
Comment (dhermes) [inline vision/google/cloud/vision/image.py]: See my comment above

(Also, trailing comma)

File: vision/unit_tests/test_annotations.py [hunk 22]
```diff
@@ -73,7 +73,7 @@ def test_from_pb(self):
         annotations = self._make_one().from_pb(image_response)
         self.assertEqual(annotations.labels, [])
         self.assertEqual(annotations.logos, [])
-        self.assertEqual(annotations.faces, ())
+        self.assertEqual(annotations.faces, [])
```
Comment (dhermes) [inline vision/unit_tests/test_annotations.py]: ??

File: vision/unit_tests/test_annotations.py [hunk 23]
```diff
@@ -107,6 +107,21 @@ def test_it(self):
         self.assertEqual(entity.locations[0].longitude, 2.0)
 
 
+class Test__make_faces_from_pb(unittest.TestCase):
+    def _call_fut(self, annotations):
+        from google.cloud.vision.annotations import _make_faces_from_pb
+        return _make_faces_from_pb(annotations)
+
+    def test_make_face_from_pb(self):
```
Comment (dhermes) [inline vision/unit_tests/test_annotations.py]: No need for a change, but this name is repetetive. Could just go with `test_it`

File: vision/google/cloud/vision/face.py (hunk 11 above)
Comment (daspecster) [inline vision/google/cloud/vision/face.py]: TL:DR;  For now I think it's needed. But depending on how we want to handle HTTP responses I may drop it in a future PR. 

2 days ago I would have said "Yeah I believe so".

But after the Speech conversation today...I could just use the value that HTTP sends back without verifying it on our side.

i.e.

HTTP
```
"landmark_type": "LEFT_EYE"
```

Protobuf
```
landmark_type: 1
```

But probably should do that in a different PR anyway.

File: vision/google/cloud/vision/image.py (hunk 14 above)
Comment (daspecster) [inline vision/google/cloud/vision/image.py]: Sorry I was vague on the kind of encoding.

The pb code base64 encodes the bytes for us. But yeah they're always bytes.

The only place it gets converted to `str` is in the comment below which is in the `to_dict` method.

File: vision/google/cloud/vision/_gax.py (hunk 18 above)
Comment (daspecster) [inline vision/google/cloud/vision/_gax.py]: Because it's handled on the constructor of [`Image`](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2922/files#diff-fd4c8b0a5b301b229c829279de2c008bR57).

File: vision/google/cloud/vision/face.py (hunk 20 above)
Comment (daspecster) [inline vision/google/cloud/vision/face.py]: Yup. [Can be seen here](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/4f5f782b0cc64920cf63a185b49a430b463d3199/vision/google/cloud/vision/likelihood.py#L21).

File: vision/unit_tests/test_annotations.py (hunk 22 above)
Comment (daspecster) [inline vision/unit_tests/test_annotations.py]: It returns a list from the [list comprehension](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2922/files#diff-764c6fd4c7ba386b0fd21683de027e6cR153) now, instead of the signature default which is `()`.

File: vision/google/cloud/vision/client.py (hunk 19 above)
Comment (daspecster) [inline vision/google/cloud/vision/client.py]: With the `_maybe_http` addition, it can go back to `None` like it should be.

File: vision/google/cloud/vision/image.py (hunk 14 above)
Comment (dhermes) [inline vision/google/cloud/vision/image.py]: Ah this is the constructor. :+1: to this change

**Comment by daspecster** [comment]
Merging after chat LGTM.
